### PR TITLE
Add initial support for keyboard modifier (and desktop support)

### PR DESCRIPTION
**Pull request by mallardduck**
So I had read about this package for Atom somewhere and was really excited to give it a shot. However I mostly use a desktop workstation and realized it won't work here. So when I saw #4 and #11 I realized that adding support for that could probably solve my problem too.

----

With this patch it's now possible to use keyboard modifies with the mouse scroll to scroll thru time.

Currently this only supports using Ctrl+Alt+Scroll, however it can likely be adapted to use any other key modifier via config. Just my first time dabbling in Atom packages so I'm not 100% on how to do that part yet. :smile:

### CODE REVIEW

**Comment by mallardduck** [comment]
Welp, actually after further testing I just realized that my toggle for keyboard modifiers doesn't seem to work at all. I had a really rough version of everything worked out a bit ago that didn't have this bug, but after refactoring somethings to use more of the original code I ended up with this.

Guess I'll give it another go and report back!

**Comment by mallardduck** [comment]
Ok, now it should, in theory, play nice! No promises that it won't eat your cat, but I've got things as ironed out as I can now. After some testing I realized that anything involving shift scroll is gonna just do horizontal; and anything control scroll is just gonna adjust zoom.

I've updated the key to be just Alt. While this does still allow the page to be scrolled as you scroll through time, atom will usually keep you around the same spot. As mentioned in the second commit message, this does work now however there still feels like room for improvement.

----

This is complete speculation and spit-balling but...maybe provide an option for operation mode; Keyboard+Mouse, or TrackPad(w/o or w/ hotkey). Then when operating in keyboard mode the listener will be attached to the keydown and keyup events; that way you can prevent the default key actions? (taking over the zooming, or side-scroll the keys cause now?) Then when those events fire you set and unset, respectively, the scroll listener.

**Comment by mallardduck** [comment]
Aye. Found a way to fix #5.

Also didn't realize it but I totally fixed #6 too as I was having a similar issue with the toggle I added. So in the process of doing this up I also fixed that and updated the name/description of that option. So it looks like in total this PR could help get rid of a bunch of these open issues. :tada: 

#4, #5, #6, #9, #10 (partially!), and #11 (since it's a repeat).

File: lib/scroll-through-time.js [hunk 1]
```diff
@@ -35,30 +43,42 @@ function getListener (direction) {
...
+      default: 'down',
       enum: [
         {value: 'left', description: 'Scroll left to undo, right to redo'},
         {value: 'right', description: 'Scroll right to undo, left to redo'},
         {value: 'up', description: 'Scroll up to undo, down to redo'},
         {value: 'down', description: 'Scroll down to undo, up to redo'},
       ],
+    },
+    keyboardState: {
+      title: 'Keyboard + Mouse Support',
```
Comment (mathieudutour) [inline lib/scroll-through-time.js]: what does "mouse" has to do with this option? It's just about having a modifier key right?

File: lib/scroll-through-time.js [hunk 2]
```diff
@@ -35,30 +43,42 @@ function getListener (direction) {
...
+      description: 'Enable support for a keyboard modifier key + scroll.',
+      type: 'boolean',
```
Comment (mathieudutour) [inline lib/scroll-through-time.js]: Could we change this do be an enum directly? Maybe `none` (default), `cmd`, `control`, `option`?

File: lib/scroll-through-time.js [hunk 3]
```diff
@@ -35,30 +43,42 @@ function getListener (direction) {
...
+      default: false
     }
   },
   subscriptions: null,
   listener: null,
-  active: false,
+  listenerKeyboard: null,
```
Comment (mathieudutour) [inline lib/scroll-through-time.js]: never used

File: lib/scroll-through-time.js [hunk 4]
```diff
@@ -69,32 +89,37 @@ export default {
...
+        if (state.active) {
+          this.setupListener()
+        }
+      }));
+
+    this.subscriptions.add(
+      atom.config.observe('scroll-through-time.keyboardState', (newValue) => {
+          this.removeListener()
+          this.listener = getListener({direction: this.getConfig('direction'), keyboard: newValue})
+          if (state.active) {
```
Comment (mathieudutour) [inline lib/scroll-through-time.js]: needs to be `this.active` otherwise it will always be the initial value

File: lib/scroll-through-time.js [hunk 5]
```diff
@@ -69,32 +89,37 @@ export default {
     this.subscriptions.add(
       atom.config.observe('scroll-through-time.direction', (newValue) => {
         this.removeListener()
-        this.listener = getListener(newValue)
-        this.setupListener()
+        this.listener = getListener({direction: newValue, keyboard: this.getConfig('keyboardState')})
+        if (state.active) {
```
Comment (mathieudutour) [inline lib/scroll-through-time.js]: needs to be `this.active` otherwise it will always be the initial value

File: lib/scroll-through-time.js [hunk 6]
```diff
@@ -35,30 +43,42 @@ function getListener (direction) {
...
   listener: null,
-  active: false,
+  listenerKeyboard: null,
 
   activate(state) {
     this.subscriptions = new CompositeDisposable();
+    this.listener = getListener({direction: state.direction, keyboard: state.keyboardState})
+
+    if (this.getConfig('active') || state.active) {
+      this.setupListener()
```
Comment (mathieudutour) [inline lib/scroll-through-time.js]: use `this.toggle(true)` so that the value of `this.active` is set as well

File: lib/scroll-through-time.js [hunk 7]
```diff
@@ -69,32 +89,37 @@ export default {
...
-    this.active = false
     this.removeListener()
     this.subscriptions.dispose();
   },
 
   serialize() {
-    return {
-      active: this.active
-    };
+    return atom.config.get('scroll-through-time');
```
Comment (mathieudutour) [inline lib/scroll-through-time.js]: hum if you do that, it means that it won't remember if it was active or not before closing atom. That's why there is `active` and `autoToggle`: it's not the same thing. `active` is the current state of activation while `autoToggle` means that it should be active on startup regardless of its previous state of activation

File: lib/scroll-through-time.js [hunk 8]
```diff
@@ -35,30 +43,42 @@ function getListener (direction) {
 export default {
 
   config: {
-    autoToggle: {
-      title: 'Auto Toggle',
-      description: 'Toggle on start.',
+    active: {
```
Comment (mathieudutour) [inline lib/scroll-through-time.js]: I don't think you should change the name of the settings, it's breaking some behaviour (see below)

File: lib/scroll-through-time.js [hunk 9]
```diff
@@ -35,30 +43,42 @@ function getListener (direction) {
...
+    keyboardState: {
+      title: 'Keyboard + Mouse Support',
+      description: 'Enable support for a keyboard modifier key + scroll.',
+      type: 'boolean',
+      default: false
     }
   },
   subscriptions: null,
   listener: null,
-  active: false,
```
Comment (mathieudutour) [inline lib/scroll-through-time.js]: shouldn't remove this. This is the current state of activation

File: lib/scroll-through-time.js (hunk 1 above)
Comment (mallardduck) [reply lib/scroll-through-time.js]: I guess I was just confused. From my initial testing of the plugin it didn't seem to work on my desktop at all. However now that I switch back to master branch it does work; I have no clue why it didn't work for me initially.

File: lib/scroll-through-time.js (hunk 3 above)
Comment (mallardduck) [reply lib/scroll-through-time.js]: Correct. Was part of the earlier patch that had a separate handler for keyboard.

File: lib/scroll-through-time.js (hunk 7 above)
Comment (mallardduck) [reply lib/scroll-through-time.js]: I guess I'm confused about the deviation of active and autoToggle in the first place?  Shouldn't the user set config values be the source of truth for the state of the the package functionality?

I admit I'm ignorant since I'm not a long term user of the package. Is the current expected workflow that users only enable/disable (toggle) the feature as needed; using the provided commands to be efficient? If that's the case I think I'm starting to get where you're coming from.

So on that same train of thought, rather than having the modifier key check in the scroll handler I should simply make the key modifier toggle the feature on press and release. It seems more inline with the current functionality I suppose.

Comment (mathieudutour) [reply lib/scroll-through-time.js]: Yes there are commands to activate or deactivate the scroll. And then there is the config which you set in the settings. One of the config is to activate the scroll on startup, regardless of the previous state.

 > I should simply make the key modifier toggle the feature on press and release

I think it's less efficient to set up the listener on each key press than having the listener and checking the key modifier.

**Comment by mallardduck** [comment]
Ok, so what I could whip up doesn't seem to play nice if we go the direction you suggested.

It's all very complex and I only have enough of a grasp to know it won't work. (at least the way I've tried) So bear with me while I try to put it into words. So we'll start with the scroll handler being used. Like what was pointed out in #5 the window still technically scrolls when you use this as is. So I adjusted the eventlistener call and invoked stopImmediatePropagation to prevent that.

While that seemed to have fixed the scrolling in the window as you scroll thru time, this didn't actually solve for the issue of preventing the default actions of Ctrl(zoom) and Alt(side scroll) when used with scroll. So to take that on we will want to throw in a preventDefault; this effectively prevents the stock scroll modifiers from kicking in.

In doing this though you end up breaking scroll everywhere in atom. So if you have scroll thru time on for your editor, you can't use scroll in your settings, etc. Now why's that though? Well as issue #9 indicates it's cause the package is active everywhere in atom and that's cause events are bound to the document instead of the specific editor element. While I claimed this PR will fix #9 what I realize is that it only fixes the symptom of it since having a modifier key would allow scroll to work in tree view without enabling time scroll.

So next I looked into doing just that, attaching the eventlistener to the editor only. At first I found success and thought it could get things on the right track, but then I realized you really only want this active in the current editor window. So then that requires even more logic track the window the user is on; remove listeners from the previous window and then add them to the new one.

So to summarize my points:

1. Any scroll listener that will use a modifier key requires to preventDevault and stopImmediatePropagation. (otherwise you're gonna be zooming or scrolling the page)
2. Doing this with the current (omnipresent, doc level listener) will prevent scroll in every area of atom.
3. Scoping the listener to just editor windows is complex; requires moving listeners as users move, or just applying it to all editor tabs (and then as more are opened. Even if it fixes some issues it clearly adds more. (also seems to make it impossible to reliably remove listeners)

**Comment by mallardduck** [comment]
Ok. So my last comment can be, more or less, ignored.

I still don't exactly like that the scroll handler applies to the whole Atom window, but now I'm just checking for the scroll context by making sure it has a parent of the atom-editor. Meaning this PR does still fix #5 and #9. (and the others I mentioned still seem fixed too!)

File: lib/scroll-through-time.js [hunk 10]
```diff
@@ -37,20 +58,32 @@ export default {
...
+      description: 'Enabled on Atom startup.',
       type: 'boolean',
       default: true
     },
     direction: {
       title: 'Scroll direction',
+      description: 'Set the scroll direction action. Must be up/down on devices without touchpad or sidescroll.',
       type: 'string',
-      default: 'left',
+      default: 'down',
```
Comment (mathieudutour) [inline lib/scroll-through-time.js]: can we keep `left` as the default?

File: lib/scroll-through-time.js [hunk 11]
```diff
@@ -2,16 +2,37 @@
...
-  let reversed = direction == 'right' || direction == 'down'
+  let reversed = options.direction == 'right' || options.direction == 'down'
+  let keyboard = options.keyboard
   let previousScroll = 0
   let timeout
   return function (e) {
-    if (Math.abs(e['delta' + mainAxis]) < Math.abs(e['delta' + crossAxis])) {
-      return
+
+      function isDescendantOfType(parentType, child) {
```
Comment (mathieudutour) [inline lib/scroll-through-time.js]: define the function on the root scope so that it doesn't need to be defined everytime

Comment (mallardduck) [reply lib/scroll-through-time.js]: Can do!

File: lib/scroll-through-time.js (hunk 10 above)
Comment (mallardduck) [reply lib/scroll-through-time.js]: If that's preferred sure thing.

However in the interest making this package more welcoming to all users I was hoping the default could be changed. Desktop users generally don't have side scrolling so for those users the default option serves no value.

Comment (mathieudutour) [reply lib/scroll-through-time.js]: I see. The issue is that by changing the default, it will break the workflow of the existing users :/

Comment (mallardduck) [reply lib/scroll-through-time.js]: Very fair point! (Didn't even think about how atom manages defaults and that changing this would affect any users with default)

**Comment by mallardduck** [comment]
No problem at all! Glad to help out with such a neat package.
It was also a great experience learning about how atom packages work. Plus helping squash bugs/issues is always fun!

:tada:
